### PR TITLE
CAF::Reporter::log(): ensure that there is no undefined argument

### LINKED ISSUE #134: CAF::Reporter::log should be protected against undefined arguments
Under some circumstances, it may happen that some of the arguments passed to `CAF;;Reporter::log()` are undef. `CAF;;Reporter::log()`  should be protected against that rather than printing a warning message...

```
#       Use of uninitialized value $_[6] in join or string at /usr/lib/perl/CAF/Reporter.pm line 367, <GEN54> line 11.
#  at /usr/lib/perl/CAF/Reporter.pm line 367
#       CAF::Reporter::log('CAF::Application=HASH(0x2a6c280)', '[VERB] ', 'Command ', 'service $service start', ' produced stdout: ', undef, ' and stderr: ', undef) called at /usr/lib/perl/CAF/Reporter.pm line 219
#       CAF::Reporter::report('CAF::Application=HASH(0x2a6c280)', '[VERB] ', 'Command ', 'service $service start', ' produced stdout: ', undef, ' and stderr: ', undef) called at /usr/lib/perl/CAF/Reporter.pm line 314
#       CAF::Reporter::verbose('CAF::Application=HASH(0x2a6c280)', 'Command ', 'service $service start', ' produced stdout: ', undef, ' and stderr: ', undef) called at /exp/si/jouvin/GitRepositories/Quattor/configuration-modules-grid/ncm-dpmlfc/target/dependency/build-scripts/Test/Quattor/Component.pm line 55
#       Test::Quattor::Component::__ANON__('NCM::Component::dpmlfc=HASH(0x2d0d8a0)', 'Command ', 'service $service start', ' produced stdout: ', undef, ' and stderr: ', undef) called at /usr/lib/perl/CAF/Object.pm line 152
#       CAF::Object::__ANON__('CAF::Service=HASH(0x35ac820)', 'Command ', 'service $service start', ' produced stdout: ', undef, ' and stderr: ', undef) called at /usr/lib/perl/CAF/Service.pm line 171
#       CAF::Service::_logcmd('CAF::Service=HASH(0x35ac820)', 'service', '$service', 'start') called at /usr/lib/perl/CAF/Service.pm line 417
#       CAF::Service::__ANON__('CAF::Service=HASH(0x35ac820)') called at /exp/si/jouvin/GitRepositories/Quattor/configuration-modules-grid/ncm-dpmlfc/target/lib/perl/NCM/Component/dpmlfc.pm line 1634
#       NCM::Component::dpmlfc::restartServices('NCM::Component::dpmlfc=HASH(0x2d0d8a0)') called at /exp/si/jouvin/GitRepositories/Quattor/configuration-modules-grid/ncm-dpmlfc/target/lib/perl/NCM/Component/dpmlfc.pm line 760
#       NCM::Component::dpmlfc::configureNode('NCM::Component::dpmlfc=HASH(0x2d0d8a0)', 'grid16.lal.in2p3.fr', 'EDG::WP4::CCM::Configuration=HASH(0x1ff5ae0)') called at src/test/perl/configure.t line 1272
```

### PR DESCRIPTION
**Pull request by jouvin**
Fixes #134

### CODE REVIEW

File: src/main/perl/Reporter.pm [hunk 1]
```diff
@@ -202,6 +202,12 @@ sub _print
 sub report
 {
     my $self = shift;
+    my $i=0;
+    # Ensure that there is no undefined arg: replace by '' if any
+    while ( $i < scalar(@_)) {
+      $_[$i] = '' if !defined($_[$i]);
```
Comment (stdweird) [inline src/main/perl/Reporter.pm]: i'm not in favour of using `''`, better to use `<undef>` or something that is obvious (as this is actually exposing a bug in whatever calls the reporter).

File: src/main/perl/Reporter.pm [hunk 2]
```diff
@@ -202,6 +202,12 @@ sub _print
 sub report
 {
     my $self = shift;
+    my $i=0;
```
Comment (stdweird) [inline src/main/perl/Reporter.pm]: better not to manipulate `@_`, use a separate variable like
```perl
my @args = map {defined($_) ? $_ : '<undef>'} @_;
```

File: src/main/perl/Reporter.pm (hunk 1 above)
Comment (jouvin) [inline src/main/perl/Reporter.pm]: I tend to disagree... See https://github.com/quattor/maven-tools/issues/74#issuecomment-180246320.

File: src/main/perl/Reporter.pm [hunk 3]
```diff
@@ -202,7 +202,9 @@ sub _print
 sub report
 {
     my $self = shift;
-    my $string = join('', @_)."\n";
+    # Ensure that there is no undefined arg: replace by '' if any
```
Comment (jrha) [inline src/main/perl/Reporter.pm]: Comment no longer matches the behaviour of the code.

**Comment by jouvin** [comment]
At the end, I decided to keep the original replacement by an empty string except if debug is set in which case `<undef>`  is used to help spotting potential problems. Do you agree?

**Comment by stdweird** [comment]
@jouvin altough i understand your point, i'm not in favour of having different output for the same message using difference debug level. even in non-debug, passing an undef to the reporter is a problem that should be dealt with. it really points to a problem in the code.

**Comment by jouvin** [comment]
I'm happy to implement what the consensus is... as long as the reporter is not raising an error. @jrha @ned21 what do you think?

**Comment by jrha** [comment]
Debug should definitely not affect the behaviour of the code.
I can't see what is wrong with always using`<undef>` personally.

**Comment by ned21** [comment]
I can see an argument for just not logging anything if undef is passed but the current implementation in 9750654 is also fine with me so I feel this can be merged.

**Pull request merged**